### DISCUSSION TITLE
Graphql-ws with apollo client v2

### DISCUSSION THREAD

**Discussion by alexsserban** [enisdenjo/graphql-ws#269] [Q&A] [answered]
Hello,

My setup consists of a backend based on apollo server, a frontend app based on Nuxt JS - which has only compatibility with Apollo client V2, and a new Frontend App with Vue3 with Apollo Client V3.

On the new app everything related to Graphql-ws works as expected, but is there a way I can make Graphql-ws work with Apollo Client v2? The recipe for apollo client from what I can see is based on apollo client v3.

Any recommendations for how I could approach the problem would be very appreciated, as I've transitioned the backend and the new front end app to graphql-ws but I also need to be able to use it in the NuxtJS app with Apollo v2.

**Comment by enisdenjo** [comment, marked as answer]
v3 is not much different from v2. Here's a quick and **untested** integration of `graphql-ws` with Apollo Client v2:

```ts
import { print } from 'graphql';
import { createClient, ClientOptions, Client } from 'graphql-ws';
import { ApolloLink, Operation, FetchResult, Observable } from 'apollo-link'; // yarn add apollo-link

// **literally** copied over from the "Client usage with Apollo" recipe
class WebSocketLink extends ApolloLink {
  private client: Client;

  constructor(options: ClientOptions) {
    super();
    this.client = createClient(options);
  }

  public request(operation: Operation): Observable<FetchResult> {
    return new Observable((sink) => {
      return this.client.subscribe<FetchResult>(
        { ...operation, query: print(operation.query) },
        {
          next: sink.next.bind(sink),
          complete: sink.complete.bind(sink),
          error: (err) => {
            if (Array.isArray(err))
              // GraphQLError[]
              return sink.error(
                new Error(err.map(({ message }) => message).join(', ')),
              );

            if (err instanceof CloseEvent)
              return sink.error(
                new Error(
                  `Socket closed with event ${err.code} ${err.reason || ''}`, // reason will be available on clean closes only
                ),
              );

            return sink.error(err);
          },
        },
      );
    });
  }
}
```

You can then use the `WebSocketLink`  as showcased in [`Apollo's v2 subscriptions documentation`](https://www.apollographql.com/docs/react/v2/data/subscriptions/) with [`graphql-ws` client options](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/client.ClientOptions.md#interface-clientoptions).

P.S. Updated the recipe to reflect the subtle differences too: https://github.com/enisdenjo/graphql-ws/commit/d9802130cd73c1f202427a1026896a9364fcef26.

**Reply by enisdenjo** [reply]
Updated the recipe to reflect the subtle differences too: https://github.com/enisdenjo/graphql-ws/commit/d9802130cd73c1f202427a1026896a9364fcef26.

**Reply by alexsserban** [reply]
Thanks for the fast reply, but unfortunately I still can't make it work and still trying to figure out the issue.

It's a [Nujs app](https://github.com/nuxt/nuxt.js) (with SSR, without TS) using [Nuxt-apollo-modules](https://github.com/nuxt-community/apollo-module) which is based on ApolloClient V2.

Queries and mutations are fine, but when I am trying to use a subscription, I get the error below. From what I can see on the backend, my subscription reaches the onConnect hook but not the onSubscribe from graphql-ws, but more than that I don't know how to debug the problem.
 
Unhandled GraphQL subscription error Error: Cannot stringify invalid message
    at stringifyMessage (graphql-ws.js?830e:170)
    at eval (graphql-ws.js?830e:537)`

```
My config looks as follows:
`import { onError } from 'apollo-link-error';
import { HttpLink } from 'apollo-link-http';
import { ApolloLink, Observable, split } from 'apollo-link';
import { getMainDefinition } from 'apollo-utilities';
import { createClient} from 'graphql-ws';

class WebSocketLink extends ApolloLink {
    client;

    constructor(options) {
        super();
        this.client = createClient(options);
    }

    request(operation) {
        return new Observable((sink) => {
            return this.client.subscribe(
                { ...operation, query: print(operation.query) },
                {
                    next: sink.next.bind(sink),
                    complete: sink.complete.bind(sink),
                    error: (err) => {
                        if (Array.isArray(err))
                            // GraphQLError[]
                            return sink.error(new Error(err.map(({ message }) => message).join(', ')));

                        if (err instanceof CloseEvent)
                            return sink.error(
                                new Error(
                                    `Socket closed with event ${err.code} ${err.reason || ''}` // reason will be available on clean closes only
                                )
                            );

                        return sink.error(err);
                    },
                }
            );
        });
    }
}

const wsLink = process.browser
    ? new WebSocketLink({
          url: 'ws://localhost:4000/graphql',
          connectionParams: () => {
              const accessToken = 'TEMP';
              return {
                  Authorization: accessToken ? `Bearer ${accessToken}` : '',
              };
          },
      })
    : null;

const httpLink = new HttpLink({
    // You should use an absolute URL here
    uri: 'http://localhost:4000/graphql',
});

const splitLink =
    wsLink !== null
        ? split(
              // split based on operation type
              ({ query }) => {
                  const definition = getMainDefinition(query);
                  return definition.kind === 'OperationDefinition' && definition.operation === 'subscription';
              },
              wsLink,
              httpLink
          )
        : httpLink;
```

**Reply by enisdenjo** [reply]
Hmm, seems like an invalid message is being sent/received. Can you please provide a screenshot of the WebSocket messages from you browsers DevTools?

If you're using Chrome, this is what you need to do:
1. Open DevTools
2. Go to "Network" tab
3. Refresh the page
4. Select the `/graphql` WS request
5. Inside the WS request drawer, select the "Frames" tab
6. Screenshot
7. Upload here

This will help me figure out who/what is the culprit, thanks in advance!

**Reply by alexsserban** [reply]
I don't seem to see any Frames Tab, only headers and message as shown below. Also I'm thinking that is something I am doing wrong in Nuxt trying to connect to graphql ws

<img width="1008" alt="Screenshot 2021-11-16 at 14 58 04" src="https://user-images.githubusercontent.com/30747926/141989496-43015e6e-bf97-47c6-82ee-a2231d1d6514.png">

<img width="512" alt="Screenshot 2021-11-16 at 14 59 25" src="https://user-images.githubusercontent.com/30747926/141989713-b323f613-c670-4bf8-b662-9abb38f61b1b.png">

**Reply by alexsserban** [reply]
The same subscription works great with the other frontend app build on Vite + Vue3, so it's definitely not a problem from the server, but a problem in the client app, either something with Nuxt, either Apollo client v2.

**Reply by enisdenjo** [reply]
The Messages tab is what I was looking for, thanks (it was named Frames some time ago, chrome probably changed it).

Nevertheless, interesting... From the looks of it, the client seems to be sending a broken message. In the 1st screenshot, on the bottom, there could be a close code + reason, can you read it out for me if there (or just give a full screenshot)?

Furthermore, please attach a `console.log` in the WebSocketLink request method. Like this:
```diff
class WebSocketLink extends ApolloLink {
    client;

    constructor(options) {
        super();
        this.client = createClient(options);
    }

    request(operation) {
        return new Observable((sink) => {
+           console.log({ ...operation, query: print(operation.query) });
            return this.client.subscribe(
                { ...operation, query: print(operation.query) },
                {
                    next: sink.next.bind(sink),
                    complete: sink.complete.bind(sink),
                    error: (err) => {
                        if (Array.isArray(err))
                            // GraphQLError[]
                            return sink.error(new Error(err.map(({ message }) => message).join(', ')));

                        if (err instanceof CloseEvent)
                            return sink.error(
                                new Error(
                                    `Socket closed with event ${err.code} ${err.reason || ''}` // reason will be available on clean closes only
                                )
                            );

                        return sink.error(err);
                    },
                }
            );
        });
    }
}
```
and share please what it reports, thanks!

**Reply by alexsserban** [reply]
Both screenshots are attached below, but I can't see any close code you were saying about. Also, another pretty weird thing is that when I reload the page, it automatically opens the browser print functionality to print the page and if I disable the subscription that doesn't happen anymore.

Also, in the console.log, the operationName is correct, that is the name of my subscription.

<img width="1010" alt="Screenshot 2021-11-16 at 16 09 21" src="https://user-images.githubusercontent.com/30747926/142000934-ac1f1310-fd4c-44fa-8c53-7d358a0c7598.png">
<img width="1016" alt="Screenshot 2021-11-16 at 16 10 14" src="https://user-images.githubusercontent.com/30747926/142000949-4c2e663c-8297-42bd-91cf-5dd62de553ff.png">

**Reply by enisdenjo** [reply]
Thats fine, no close event because the error is scoped to a single subscription, hence not fatal (just wanted to confirm).

However, I see the problem, notice the operation `query` field - it is `undefined` - this is why the client is failing. The `query` field is required and must be always present, please ~~check your Apollo implementation~~ see https://github.com/enisdenjo/graphql-ws/discussions/269#discussioncomment-1651783. 😄

**Reply by enisdenjo** [reply]
@AlexandruSerban142 I figured out the exact problem, looking at your config over at: https://github.com/enisdenjo/graphql-ws/discussions/269#discussioncomment-1651032, you dont seem to import the `print` function from `graphql`, you're using the [browser's window `print`](https://developer.mozilla.org/en-US/docs/Web/API/Window/print) instead haha (explains the print dialog too).

By importing `print` from `graphql`, you should be good to go.
```diff
import { onError } from 'apollo-link-error';
import { HttpLink } from 'apollo-link-http';
import { ApolloLink, Observable, split } from 'apollo-link';
import { getMainDefinition } from 'apollo-utilities';
import { createClient} from 'graphql-ws';
+ import { print } from 'graphql';

class WebSocketLink extends ApolloLink {
    client;

    constructor(options) {
        super();

...
```

**Reply by alexsserban** [reply]
Hahah, yes, now everything looks good... Thanks a lot for all the help.
